### PR TITLE
Implement pg_get_constraintdef

### PR DESCRIPTION
**Pull request by guswynn**
### Motivation

* This PR adds a known-desirable feature: #9483

Note that this cannot get the relevant information from the empty-in-materialize `pg_constraints` table, so it alway's error's when actually called with data.
Also the new `UnaryFunc::NotImplemented` just takes in a full `String`, it could be switched to `Cow<'static, str>` for a slight optimization if need be

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: misc/python/materialize/cli/mzcompose.py [hunk 1]
```diff
@@ -87,9 +87,11 @@ def main(argv: List[str]) -> int:
 
     # From here on out we're definitely invoking Docker Compose, so make sure
     # it's new enough.
-    output = spawn.capture(
-        ["docker-compose", "version", "--short"], unicode=True
-    ).strip()
+    output = (
```
Comment (quodlibetor) [inline misc/python/materialize/cli/mzcompose.py:90]: this should go into a separate PR.

File: src/expr/src/scalar/func.rs [hunk 2]
```diff
@@ -2115,6 +2115,7 @@ fn jsonb_pretty<'a>(a: Datum<'a>, temp_storage: &'a RowArena) -> Datum<'a> {
     Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzEnumReflect,
 )]
 pub enum BinaryFunc {
+    NotImplemented(String),
```
Comment (quodlibetor) [inline src/expr/src/scalar/func.rs:2118]: could you either add a comment explaining what the string is or change it to a self-documenting item?

```suggestion
    NotImplemented { not_implemented_item: String },
```

File: src/sql/src/func.rs [hunk 3]
```diff
@@ -1610,6 +1610,15 @@ lazy_static! {
...
                 params!(Int64) => sql_impl_func("CASE WHEN $1 = 6 THEN 'UTF8' ELSE NULL END"), 1597;
             },
+            // pg_get_constraintdef gives more info about a constraint with in the `pg_constraint`
+            // view. It currently returns no information as the `pg_constraint` view is empty in
+            // materialize
+            "pg_get_constraintdef" => Scalar {
+                params!(Oid) => {
+                    UnaryFunc::NotImplemented("pg_get_constraintdef".to_string())}, 1598;
+                params!(Oid, Bool) => {
+                    BinaryFunc::NotImplemented("pg_get_constraintdef".to_string())}, 1599;
```
Comment (quodlibetor) [inline src/sql/src/func.rs:1620]: how did you choose these OIDs? I'm not sure how we choose them.

File: test/testdrive/pg_get_constraintdef.td [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+# Without EXTENDED, SHOW TYPES shouldn't have anything to return.
+
+! SELECT pg_get_constraintdef(885)
+pg_get_constraintdef is not yet implemented is undefined
```
Comment (quodlibetor) [inline test/testdrive/pg_get_constraintdef.td:13]: Ah it seems like we should at a `NotImplemented` variant to the error enum instead of reusing `::Undefined`, that would get rid of this `is undefined`

**Review by quodlibetor — CHANGES_REQUESTED** [review]
This looks amazing! Much more than we expected. I'd definitely like Matt to do final review, but I've got a couple comments.

File: misc/python/materialize/cli/mzcompose.py (hunk 1 above)
Comment (guswynn) [reply misc/python/materialize/cli/mzcompose.py:90]: will split it out!

File: src/sql/src/func.rs (hunk 3 above)
Comment (guswynn) [reply src/sql/src/func.rs:1620]: I'm not sure what you mean

File: test/testdrive/pg_get_constraintdef.td (hunk 4 above)
Comment (guswynn) [reply test/testdrive/pg_get_constraintdef.td:13]: ah yeah, will do that!

File: test/sqllogictest/pg_get_constraintdef.slt [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+#
+# The original source code is subject to the terms of the Apache
+# 2.0 license, a copy of which can be found in the LICENSE file at the
+# root of this repository.
+
+query T
+SELECT pg_get_constraintdef(pg_constraint.oid) FROM pg_constraint
+----
+
+query error
```
Comment (mjibson) [inline test/sqllogictest/pg_get_constraintdef.slt:21]: This should have more text after error to match the error string on (it's a regex, and can match a part of the error text).

File: test/testdrive/pg_get_constraintdef.td [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+# Without EXTENDED, SHOW TYPES shouldn't have anything to return.
+
+! SELECT pg_get_constraintdef(885)
```
Comment (mjibson) [inline test/testdrive/pg_get_constraintdef.td:12]: I'm not convinced this needs to be in both td and slt. I think just the slt tests are sufficient.

File: src/sql/src/func.rs [hunk 7]
```diff
@@ -1610,6 +1610,15 @@ lazy_static! {
                 // encoding id for UTF8 (6) is provided, otherwise return 'NULL'.
                 params!(Int64) => sql_impl_func("CASE WHEN $1 = 6 THEN 'UTF8' ELSE NULL END"), 1597;
             },
+            // pg_get_constraintdef gives more info about a constraint with in the `pg_constraint`
+            // view. It currently returns no information as the `pg_constraint` view is empty in
+            // materialize
+            "pg_get_constraintdef" => Scalar {
+                params!(Oid) => {
```
Comment (mjibson) [inline src/sql/src/func.rs:1617]: Don't need to add a new func variant here, it's ok to just return an error in a closure. See https://github.com/MaterializeInc/materialize/blob/23c4aff5caf3b4a66abaa0f71f73b3f35a45e860/src/sql/src/func.rs#L1339 for an example.

File: src/sql/src/func.rs (hunk 3 above)
Comment (mjibson) [reply src/sql/src/func.rs:1620]: Oids are unique identifiers in postgres. Every thing has one and they are globally unique: tables, types, operators, functions, etc. Here's the correct oids:

```
postgres=# select oid, proname, proargtypes from pg_proc where proname like 'pg_get_constraint%';
 oid  |       proname        | proargtypes 
------+----------------------+-------------
 1387 | pg_get_constraintdef | 26
 2508 | pg_get_constraintdef | 26 16
```

So 1387 for the single arg, and 2508 for the two argument.

File: src/sql/src/func.rs (hunk 7 above)
Comment (guswynn) [reply src/sql/src/func.rs:1617]: TIL about `bail_unsupported`!

**Comment by guswynn** [comment]
@mjibson @quodlibetor updated with comments! Note that with the `bail_unsupported!` change, all instances of this query immediately error

File: test/sqllogictest/pg_get_constraintdef.slt [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+#
+# This file is derived from the logic test suite in CockroachDB. The
+# original file was retrieved on June 10, 2019 from:
+#
+# The original source code is subject to the terms of the Apache
+# 2.0 license, a copy of which can be found in the LICENSE file at the
+# root of this repository.
+
+query error pg_get_constraintdef not yet supported
+SELECT pg_get_constraintdef(pg_constraint.oid) FROM pg_constraint
```
Comment (mjibson) [inline test/sqllogictest/pg_get_constraintdef.slt:18]: Hmm, this is unfortunately not what we need. We need a query with this to successfully run, just that it won't ever have any rows so will happen to never call this function. We might need to error at a deeper level.

Comment (guswynn) [reply test/sqllogictest/pg_get_constraintdef.slt:18]: the previous version of this commit had that behavior, with UnaryFunc::NotImplemented. does that work?

Comment (guswynn) [reply test/sqllogictest/pg_get_constraintdef.slt:18]: I don't think so, as far as I can tell `NullaryFunc`'s must be infallible

File: src/expr/src/scalar/func.rs [hunk 9]
```diff
@@ -3999,6 +4006,7 @@ impl UnaryFunc {
             | TrimTrailingWhitespace
             | Upper
             | Lower => false,
+            NotImplemented { .. } => true,
```
Comment (mjibson) [inline src/expr/src/scalar/func.rs:4009]: this probably needs to be false?

File: src/expr/src/scalar/func.rs [hunk 10]
```diff
@@ -4200,6 +4208,7 @@ impl UnaryFunc {
             | CastStringToVarChar(_)
             | CastCharToString(_)
             | CastFloat32ToFloat64(_) => unreachable!(),
+            NotImplemented { function_name } => f.write_str(function_name),
```
Comment (mjibson) [inline src/expr/src/scalar/func.rs:4211]: I don't have a good understanding of what this method does, but can you try dropping it into the unreachable block above?

**Review by mjibson — APPROVED** [review]

**Review by quodlibetor — APPROVED** [review]
🔥
